include/ensmallen_bits/moead/moead.hpp

```diff
@@ -0,0 +1,316 @@
+/**
+ * @file moead.hpp
+ * @author Utkarsh Rai
+ * @author Nanubala Gnana Sai
+ *
+ * MOEA/D, Multi Objective Evolutionary Algorithm based on Decompositon is a
+ * multi objective optimization algorithm. It employs evolutionary algorithms,
+ * to find better solutions by iterating on the previous solutions and
+ * decomposition approaches, to convert the multi objective problem to a single
+ * objective one, to find the best Pareto Front for the given problem.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_MOEAD_MOEAD_HPP
+#define ENSMALLEN_MOEAD_MOEAD_HPP
+
+namespace ens {
+
+/**
+ * This class implements the MOEA/D algorithm with Differential Evolution
+ * crossover. Step numbers used in different parts of the implementation
+ * correspond to the step number used in the original algorithm by the author.
+ *
+ * For more information, see the following:
+ * @code
+ * @article{article,
+ * author = {Zhang, Qingfu and Li, Hui},
+ * year = {2008},
+ * pages = {712 - 731},
+ * title = {MOEA/D: A Multiobjective Evolutionary Algorithm Based on
+ *    Decomposition},
+ * journal = {Evolutionary Computation, IEEE Transactions on},
+ *
+ * @article{4633340,
+ * author={H. {Li} and Q. {Zhang}},
+ * year={2009},
+ * pages={284-302},}
+ * title={Multiobjective Optimization Problems With Complicated Pareto Sets, MOEA/D and NSGA-II},
+ * journal={IEEE Transactions on Evolutionary Computation},
+ * @endcode
+ *
+ * MOEA/D can optimize arbitrary multi-objective functions. For more details,
+ * see the documentation on function types included with this distribution or
+ * on the ensmallen website.
+ */
+class MOEAD {
+ public:
+  /**
+   * Constructor for the MOEA/D optimizer.
+   *
+   * The default values provided here are not necessarily suitable for a
+   * given function. Therefore, it is highly recommended to adjust the
+   * parameters according to the problem.
+   *
+   * @param populationSize The number of elements in the population.
+   * @param maxGenerations The maximum number of generations allowed.
+   * @param crossoverProb The probability that a crossover will occur.
+   * @param neighborSize The number of nearest neighbours of weights
```

Comment:
Maybe it makes sense to have `neighborProb` first followed by `neighborSize` instead of putting `distributionIndex` in between.